perfkitbenchmarker/linux_packages/ntp.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2016 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module containing NTP installation and cleanup functions."""
+
+import logging
+import re
+
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import vm_util
+
+
+def YumInstall(vm):
+  """Installs the NTP package on the VM."""
+  vm.InstallPackages('ntp')
+
+
+def AptInstall(vm):
+  """Installs the NTP package on the VM."""
+  vm.InstallPackages('ntp')
+
+
+def IsNTPSynced(vm):
+  """Determine whether a VM's clock is NTP synchronized.
+
+  Args:
+    vm: a virtual_machine.VirtualMachine object.
+
+  Returns:
+    bool. True if clock is synchronized by NTP, False if not.
+  """
+
+  # See https://www.eecis.udel.edu/~mills/ntp/html/decode.html#sys for
+  # why sync_ntp is the right string to search for. Note that we test
+  # that the VM's clock is synchronized specifically by NTP, so if it
+  # syncs via some special-purpose clock hardware, we will return
+  # False.
+  out, err = vm.RemoteCommand('ntpq -c "readvar 0"', should_log=True)
+
+  if re.search('sync_ntp', out) is not None:
+    return True
+  else:
+    return False
+
+
+@vm_util.Retry()
+def WaitForNTPSync(vm):
+  """Wait until the given VM's clock is synced via NTP.
+
+  Args:
+    vm: a virtual_machine.VirtualMachine object.
+
+  Raises:
+    errors.Resource.RetryableCreationError if the wait time exceeds a deadline.
+
+  Returns:
+    None.
+  """
+  if not IsNTPSynced(vm):
+    raise errors.Resource.RetryableCreationError('NTP has not yet synced.')
+
+
+@vm_util.Retry()
+def WaitForNTPDaemon(vm):
+  """Wait until the NTP daemon has started on the given VM."""
+
+  vm.RemoteCommand('sudo service ntp start', should_log=True)
+
+  out, err = vm.RemoteCommand('pgrep ntpd')
```

Comment:
Changed to use `--exact`, and account for pgrep's non-zero exit. I don't want to use `service` because I don't know if that will change in future systemd-based Ubuntus.